### ISSUE TITLE
Encoded hash is not being decoded (rounding bug)

### ISSUE THREAD

**Issue by MarkOtter** [ullmark/hashids.net#23]
I have a reproducable bug in which certain input does create a valid hashid, but the Decode fails!

Repro code where the result is empty, instead of equal to the source:

````C#
var hashId = new Hashids(salt: "0Q6wKupsoahWD5le", alphabet: "abcdefghijklmnopqrstuvwxyz1234567890!", seps: "cfhistu");

var source = new long[] { 35887507618889472L, 30720L, Int64.MaxValue };

var encoded = hashId.EncodeLong(source);

var result = hashId.DecodeLong(encoded);

// ASSERT result == source FAILS
`````
_(Both the first (`35887507618889472L`) and the last (`Int64.MaxValue`) long values fail here)_

After some investigation of my own, it turns out the root cause is the use of the `Math.Pow` function in the Unhash function. Due to its `double` nature, it is subject to rounding errors.

Proposed solution: update the `Unhash` function with the following code:

````C#
private long Unhash(string input, string alphabet)
{
    long number = 0;
    var alphabetLength = new System.Numerics.BigInteger(alphabet.Length);

    for (var i = 0; i < input.Length; i++)
    {
        var pos = (long)alphabet.IndexOf(input[i]);
        number += (pos * (long)System.Numerics.BigInteger.Pow(alphabetLength, input.Length - i - 1));
    }

    return number;
}
````

**Comment by ullmark** [comment]
Cool, thanks! I'll have a look and include it in the next version!

**Comment by douglasg14b** [comment]
@MarkOtter Maybe make a pull request to get the fix in? It doens't seem like this project has been updates for a few years.

**Comment by minhhungit** [comment]
@ullmark @douglasg14b hi any news sir

have you had a chance to review this issue @ullmark

**Comment by ullmark** [comment]
A pull request would have been nice :) I'll update the code and make sure tests execute and then make a release.

**Comment by ullmark** [comment]
Doesn't seem "System.Numerics" is included in .netstandard 1.0...

**Comment by ullmark** [comment]
I could apply @MarkOtter when not using .netstandard1.0. Would that suffice @minhhungit ?

**Comment by minhhungit** [comment]
I'm still using .net framework so I don't worry much about .netstandard

**Comment by MarkOtter** [comment]
It seems like System.Numerics is available in [Standard 2.0+](https://apisof.net/catalog/System.Numerics), or there is a NuGet package for [Standard 1.1+](https://www.nuget.org/packages/System.Runtime.Numerics/).

@ullmark I think the decision should be made what minimum version you'd like to support with this code. Given the [implementation support](https://docs.microsoft.com/en-us/dotnet/standard/net-standard) chart, I think going with at least 1.1 is a safe decision.

**Comment by DwayneBull** [comment]
Given that the powers are small ( only as long as the hashed string ) it would be better to use a manual pow function and avoid bigint.

i.e.
` 
        private long Unhash(string input, string alphabet)
        {
            long number = 0;

            for (var i = 0; i < input.Length; i++)
            {
                var pos = alphabet.IndexOf(input[i]);
                number += (long)(pos * Pow(alphabet.Length, input.Length - i - 1));
            }

            return number;
        }

 private static long Pow(int target, int power)
    {
        if (power == 0) return 1;
        long result = target;
        while (power > 1)
        {
            result *= target;
            power--;
        }
        return result;
    }`
